src/RegistrationController.js

```diff
@@ -61,86 +65,141 @@ function (
+      }
+    },
+    registerUser: function (postData) {
+      var self = this;
+      this.model.attributes = postData;
+      Backbone.Model.prototype.save.call(this.model).then(function() {
+        var activationToken = self.model.get('activationToken');
+        var postSubmitData = activationToken ? activationToken : self.model.get('userName');
+        self.settings.postSubmit(postSubmitData, function() {
+          self.doPostSubmit();
+        }, function(errors) {
+          self.showErrors(errors);
         });
-        form.add(requiredFieldsLabel);
-        this.add(form);
-        this.footer = new this.Footer(this.toJSON());
-        this.add(this.footer);
-        this.addListeners();
-      }, this);
+      }).fail(function(err) {
+        var responseJSON = err.responseJSON;
+        if (responseJSON && responseJSON.errorCauses.length) {
+          var errMsg = responseJSON.errorCauses[0].errorSummary;
+          self.settings.callGlobalError(new Errors.RegistrationError(errMsg));
+        }
+      });
+    },
+    createRegistrationModel: function (modelProperties) {
+      var self = this;
+      var Model = Okta.Model.extend({
+        url: self.getRegistrationApiUrl()+'/register',
+        settings: this.settings,
+        appState: this.options.appState,
+        props: modelProperties,
+        local: {
+          activationToken: 'string'
+        },
+        toJSON: function() {
+          var data = Okta.Model.prototype.toJSON.apply(this, arguments);
+          return {userProfile: data};
+        },
+        parse: function(resp) {
+          this.set('activationToken', resp.activationToken);
+          delete resp.activationToken;
+          return resp;
+        },
+        save: function() {
+          this.settings.preSubmit(this.attributes, function(postData){
+            self.registerUser(postData);
+          }, function(errors) {
+            self.showErrors(errors);
+          });
+        }
+      });
+      return new Model();
+    },
+    showErrors: function (error, hideRegisterButton) {
+      //for parseSchema error hide form and show error at form level
+      if(error.callback == 'parseSchema' && error.errorCauses) {
+        error.errorSummary = _.clone(error.errorCauses[0].errorSummary);
+        delete error.errorCauses;
+      }
+      //show error on form
+      this.model.trigger('error', this.model, {
+        responseJSON: error
+      });
+
+      //throw global error
+      var errMsg = error.callback? error.callback+':'+ error.errorSummary: error.errorSummary;
+      this.settings.callGlobalError(new Errors.RegistrationError(errMsg));
 
-      schema.fetch().then(function(){
-        postSchemaFetch();
+      if (hideRegisterButton) {
+        this.$el.find('.button-primary').hide();
+      }
+    },
+    fetchInitialData: function () {
+      var self = this;
+      // register parse complete event listener
+      self.state.get('schema').on('parseComplete', function(updatedSchema) {
```

Comment:
I think this is probably okay, but I was actually thinking that it would look more like:

- fetchInitialData returns a new promise that is only resolved when parseComplete is finished. It's currently resolved when the schema fetch is done, which won't account for any async logic a consumer adds in their parseSchema hook
- In a preRender function, we can then go ahead and use the existing schema (at this time, we know that it will be ready) to go ahead and do this logic to add all the form elements and fields.

Overall idea being that fetchInitialData handles all async data loading (and parsing), and then we can just do view logic like adding elements, etc in a more normal synchronous way like we do elsewhere.